applications/sintering/include/pf-applications/sintering/tools.h

```diff
@@ -144,19 +151,102 @@ namespace Sintering
+        {
+          std::vector<unsigned int> refinements(subdivisions.size());
+
+          for (unsigned int d = 0; d < subdivisions.size(); ++d)
+            {
+              const auto pair =
+                decompose_to_prime_tuple(subdivisions[d], max_prime);
+
+              subdivisions[d] = pair.first;
+              refinements[d]  = pair.second;
+            }
+
+          n_refinements_base =
+            *std::min_element(refinements.begin(), refinements.end());
+
+          for (unsigned int d = 0; d < subdivisions.size(); ++d)
+            subdivisions[d] *= std::pow(2, refinements[d] - n_refinements_base);
+        }
+
+      return n_refinements_base;
+    }
+
+    template <int dim, typename Triangulation>
+    void
+    impose_periodicity(Triangulation &tria)
+    {
+      // Need to work with triangulation here
+      std::vector<
+        GridTools::PeriodicFacePair<typename Triangulation::cell_iterator>>
+        periodicity_vector;
+
+      for (unsigned int d = 0; d < dim; ++d)
+        {
+          GridTools::collect_periodic_faces(
+            tria, 2 * d, 2 * d + 1, d, periodicity_vector);
+        }
+
+      tria.add_periodicity(periodicity_vector);
+    }
+
+    template <typename Triangulation>
+    std::pair<unsigned int, unsigned int>
+    make_initial_refines(Triangulation &     tria,
+                         const InitialRefine refine,
+                         const unsigned int  n_refinements_base,
+                         const unsigned int  n_refinements_interface)
+    {
+      unsigned int n_global  = 0;
+      unsigned int n_delayed = 0;
+      if (refine == InitialRefine::Base)
+        {
+          tria.refine_global(n_refinements_base);
+
+          n_global  = n_refinements_base;
+          n_delayed = n_refinements_interface;
+        }
+      else if (refine == InitialRefine::Full)
+        {
+          tria.refine_global(n_refinements_base + n_refinements_interface);
+
+          n_global  = n_refinements_base + n_refinements_interface;
+          n_delayed = 0;
+        }
+      else
+        {
+          n_global  = 0;
+          n_delayed = n_refinements_base + n_refinements_interface;
+        }
+
+      return {n_global, n_delayed};
+    }
+  } // namespace internal
+
   template <typename Triangulation, int dim>
   unsigned int
-  create_mesh(Triangulation &     tria,
-              const Point<dim> &  bottom_left,
-              const Point<dim> &  top_right,
-              const double        interface_width,
-              const double        divisions_per_interface,
-              const bool          periodic,
-              const InitialRefine refine,
-              const unsigned int  max_prime                          = 0,
-              const double        max_level0_divisions_per_interface = 1.0,
-              const unsigned int  divisions_per_element              = 1,
-              const bool          print_stats                        = true)
+  create_mesh_from_interface(
```

Comment:
Could you add a comment describing what each function does?

Reply:
Yep, done!